### ISSUE TITLE
Completion has extraneous or broken code

### ISSUE THREAD

**Issue by mainej** [clojure-lsp/clojure-lsp#810]
This more of a question than a bug. On [these lines](https://github.com/clojure-lsp/clojure-lsp/blob/99db559557aa6385c26ce0b417cc8fc281ff283d/lib/src/clojure_lsp/feature/completion.clj#L382-L386), the `try-column` loop variable isn't used in `q/find-element-under-cursor`. So, either the loop is extraneous, or some feature is broken. There aren't any tests that would explain what the feature might be. It looks like `try-column` was used at one point, but was lost in this refactoring: c1ffd01f497aea804f124df2453805cc5a082a03. @ericdallo do you know what this code is for?

**Comment by ericdallo** [comment]
Yes, we should use the `try-column` on the find-element-under-cursor call, basically what this code does which is quite important is:
when you complete something and there is no element on the current row / col, we try with the current row / col-1 until we find a element, example:
```clojure
foo |
```
We will find nothing on `|` but on the next loop find `foo`, does that makes sense? we are probably missing a test for that

**Comment by mainej** [comment]
At first your explanation made sense, but now I'm having second thoughts...

**tldr**: I still think we should remove the loop.

To build on your example, suppose the code looks like this:
```clojure
(def foobar 1)

foo |
```
I think you're saying completion should offer `foobar` and should resolve like so:
```clojure
(def foobar 1)

foobar|
```
But, there would be a problem—maybe it's not what you want. Perhaps because you're on whitespace you want to be shown `foobar`, plus all 899 core Clojure/Java items. And regardless of what you choose, you want it to be inserted after the cursor. If you choose `foobar`, you want it to resolve like so:
```clojure
(def foobar 1)

foo foobar|
```
It's ambiguous, but I believe the editors expect this second interpretation. And that's good, because it's what the code implements currently.

So, how does this relate to `try-column`? I think the thing to remember is that `try-column` affects `cursor-element` (the analysis element). But `cursor-element` doesn't decide what's getting completed. That's decided by `cursor-loc`, the zipper location. If you were to move `cursor-loc` back too, you could make the code follow the first interpretation, resolving `foo` to `foobar`.

But moving only `cursor-element` back has little impact. `cursor-element` has two purposes:

1. It helps exclude the current item from completion. For example, `:bazz|` (notice, no space) doesn't offer `:bazz` as a completion.
2. It helps figure out the true namespace of an aliased keyword, so that completions from that namespace can be offered.

How would using `try-column` affect these two things?

1. `:bazz |` (notice the extra space) would _not_ offer `:bazz` as a completion.
2. `(ns a (:require [bee :as b])) ::b/bumble |` _would_ offer `::b/bumble-bee` as a completion.

These two effects are tiny, so it really doesn't matter much. But, IMO, both would make the behavior worse, not better. In the first case maybe you want `:bazz :bazz` and so excluding `:bazz` would be annoying.  And in the second case, you're on whitespace... why should the code assume you want to complete a keyword from the `b` alias?

And now that I've spent way too much time thinking about these five lines of code... :) I'll leave it up to you. I'm happy to fix or remove the loop. I vote for removing it.

**Comment by ericdallo** [comment]
@mainej you are right, I recall why we introduced that, that was before we fix the completion of invalid clojure code like `foo/`:
1. User has typed this code:
```clojure
(ns foo (:require [foobar :as foo]))

foo|
```
At this time, we have correct clj-kondo analysis, but when user types the `/` and becomes `foo/`, we were sending a invalid clj-kondo code which resulted in invalid analysis and then not updating analysis, then the workaround we did was to check previous columns to find a `element` and then workaround on clojure-lsp side.

Now this was already fixed on both clj-kondo and clojure-lsp and indeed I don't think we need that loop anymore :)

**Comment by mainej** [comment]
Ah, that makes perfect sense! I'll remove the loop.
